### ISSUE TITLE
Error with PerformanceCounter

### ISSUE THREAD

**Issue by antoniogarcez** [microsoft/ApplicationInsights-dotnet#2188]
Hello guys,

I updated my Application Insights' nugets of an application that I developed with .NET Framework 4.8. I installed the lastest version (2.17.0) and when I am trying to debug my application, it always gives the same exception (`'System.NullReferenceException' in Microsoft.AI.PerfCounterCollector.dll`) on this line `performanceCounter.Collect();`.

What can I do to fix this?

**Comment by cijothomas** [comment]
Please share a minimal repro app showing this issue to investigate.

**Comment by antoniogarcez** [comment]
Hi @cijothomas ,

I can show you this image when exception is thrown:

![](https://user-images.githubusercontent.com/5618128/112511037-efe91d00-8d89-11eb-8fac-59d1cb950c22.png)

And this is the output created by debug:

`Application Insights Telemetry (unconfigured): {"name":"AppDependencies","time":"2021-03-25T16:44:34.7570189Z","tags":{"ai.cloud.roleInstance":"CPU-244.famo.pt","ai.operation.id":"e1300e0f7e6464499a12a7ea536bd46a","ai.operation.parentId":"fee080fa494db64f","ai.operation.name":"GET /FAMO.CODE_API/api/Products/Lines/Configurator","ai.location.ip":"::1","ai.internal.sdkVersion":"rddf:2.17.0-146","ai.internal.nodeName":"CPU-244.famo.pt"},"data":{"baseType":"RemoteDependencyData","baseData":{"ver":2,"name":"URANO | CODE_DEV","id":"780dd643525be04f","duration":"00:00:00.0053963","success":true,"type":"SQL","target":"URANO | CODE_DEV","properties":{"DeveloperMode":"true","_MS.ProcessedByMetricExtractors":"(Name:'Dependencies', Ver:'1.1')"}}}}`  
`Application Insights Telemetry (unconfigured): {"name":"AppDependencies","time":"2021-03-25T16:44:34.7648878Z","tags":{"ai.cloud.roleInstance":"CPU-244.famo.pt","ai.operation.id":"e1300e0f7e6464499a12a7ea536bd46a","ai.operation.parentId":"fee080fa494db64f","ai.operation.name":"GET /FAMO.CODE_API/api/Products/Lines/Configurator","ai.location.ip":"::1","ai.internal.sdkVersion":"rddf:2.17.0-146","ai.internal.nodeName":"CPU-244.famo.pt"},"data":{"baseType":"RemoteDependencyData","baseData":{"ver":2,"name":"URANO | CODE_DEV","id":"5d5620c51499424e","duration":"00:00:00.0076206","success":true,"type":"SQL","target":"URANO | CODE_DEV","properties":{"DeveloperMode":"true","_MS.ProcessedByMetricExtractors":"(Name:'Dependencies', Ver:'1.1')"}}}}`  
`Exception thrown: 'System.IO.IOException' in mscorlib.dll`  
`Application Insights Telemetry (unconfigured): {"name":"AppDependencies","time":"2021-03-25T16:44:34.7759208Z","tags":{"ai.cloud.roleInstance":"CPU-244.famo.pt","ai.operation.id":"e1300e0f7e6464499a12a7ea536bd46a","ai.operation.parentId":"fee080fa494db64f","ai.operation.name":"GET /FAMO.CODE_API/api/Products/Lines/Configurator","ai.location.ip":"::1","ai.internal.sdkVersion":"rddf:2.17.0-146","ai.internal.nodeName":"CPU-244.famo.pt"},"data":{"baseType":"RemoteDependencyData","baseData":{"ver":2,"name":"URANO | CODE_DEV","id":"f41e9031a6a7584c","duration":"00:00:00.0073421","success":true,"type":"SQL","target":"URANO | CODE_DEV","properties":{"DeveloperMode":"true","_MS.ProcessedByMetricExtractors":"(Name:'Dependencies', Ver:'1.1')"}}}}`  
`Application Insights Telemetry (unconfigured): {"name":"AppDependencies","time":"2021-03-25T16:44:34.7914830Z","tags":{"ai.cloud.roleInstance":"CPU-244.famo.pt","ai.operation.id":"e1300e0f7e6464499a12a7ea536bd46a","ai.operation.parentId":"fee080fa494db64f","ai.operation.name":"GET /FAMO.CODE_API/api/Products/Lines/Configurator","ai.location.ip":"::1","ai.internal.sdkVersion":"rddf:2.17.0-146","ai.internal.nodeName":"CPU-244.famo.pt"},"data":{"baseType":"RemoteDependencyData","baseData":{"ver":2,"name":"URANO | CODE_DEV","id":"1dc255805da0d749","duration":"00:00:00.0044607","success":true,"type":"SQL","target":"URANO | CODE_DEV","properties":{"DeveloperMode":"true","_MS.ProcessedByMetricExtractors":"(Name:'Dependencies', Ver:'1.1')"}}}}`  
`Application Insights Telemetry (unconfigured): {"name":"AppDependencies","time":"2021-03-25T16:44:34.7986851Z","tags":{"ai.cloud.roleInstance":"CPU-244.famo.pt","ai.operation.id":"e1300e0f7e6464499a12a7ea536bd46a","ai.operation.parentId":"fee080fa494db64f","ai.operation.name":"GET /FAMO.CODE_API/api/Products/Lines/Configurator","ai.location.ip":"::1","ai.internal.sdkVersion":"rddf:2.17.0-146","ai.internal.nodeName":"CPU-244.famo.pt"},"data":{"baseType":"RemoteDependencyData","baseData":{"ver":2,"name":"URANO | CODE_DEV","id":"171a6434f6990746","duration":"00:00:00.0055291","success":true,"type":"SQL","target":"URANO | CODE_DEV","properties":{"DeveloperMode":"true","_MS.ProcessedByMetricExtractors":"(Name:'Dependencies', Ver:'1.1')"}}}}`  
`Application Insights Telemetry (unconfigured): {"name":"AppDependencies","time":"2021-03-25T16:44:34.8058908Z","tags":{"ai.cloud.roleInstance":"CPU-244.famo.pt","ai.operation.id":"e1300e0f7e6464499a12a7ea536bd46a","ai.operation.parentId":"fee080fa494db64f","ai.operation.name":"GET /FAMO.CODE_API/api/Products/Lines/Configurator","ai.location.ip":"::1","ai.internal.sdkVersion":"rddf:2.17.0-146","ai.internal.nodeName":"CPU-244.famo.pt"},"data":{"baseType":"RemoteDependencyData","baseData":{"ver":2,"name":"URANO | CODE_DEV","id":"fafa747f291ba04f","duration":"00:00:00.0045239","success":true,"type":"SQL","target":"URANO | CODE_DEV","properties":{"DeveloperMode":"true","_MS.ProcessedByMetricExtractors":"(Name:'Dependencies', Ver:'1.1')"}}}}`  
`Exception thrown: 'System.InvalidOperationException' in System.dll`  
`Exception thrown: 'System.NullReferenceException' in Microsoft.AI.PerfCounterCollector.dll`  
`Object reference not set to an instance of an object.`

**Comment by cijothomas** [comment]
This doesn't help us with any investigation. Need to have a repro app, or the exact instructions which can repro the behavior.

**Comment by antoniogarcez** [comment]
Sorry if it is not enough. In my MVC project, I only click in "IIS" and before my application shows up, this exception is thrown multiple times.

**Comment by Robert-Zandberg** [comment]
@antoniogarcez  Did you find the solution for your problem? I'm facing the same issue. 
I'm running an web application ASP.NET 4.6.1 I'm trying to add a default Application Insights setup by Configuring a default App Application insights by SDK.  

[ApplicationInsightsConfigFile.zip](https://github.com/microsoft/ApplicationInsights-dotnet/files/6733747/ApplicationInsightsConfigFile.zip)


Debug output -- 
`Application Insights Telemetry: {"name":"AppMetrics","time":"2021-06-29T13:19:11.0000000Z","iKey":"xxxxx","tags":{"ai.cloud.roleInstance":"LAPTOP-XJEA31","ai.internal.sdkVersion":"m-agg2:2.17.0-146"},"data":{"baseType":"MetricData","baseData":{"ver":2,"metrics":[{"name":"Server response time","kind":"Aggregation","value":572.0885,"count":1,"min":572.0885,"max":572.0885,"stdDev":0}],"properties":{"operation/synthetic":"False","request/performanceBucket":"500ms-1sec","cloud/roleName":"Unknown","_MS.MetricId":"requests/duration","_MS.IsAutocollected":"True","_MS.AggregationIntervalMs":"50000","Request.Success":"True","cloud/roleInstance":"LAPTOP-XJEA31","request/resultCode":"200"}}}}
Application Insights Telemetry: {"name":"AppMetrics","time":"2021-06-29T13:19:11.0000000Z","iKey":"xxxxx","tags":{"ai.cloud.roleInstance":"LAPTOP-XJEA31","ai.internal.sdkVersion":"m-agg2:2.17.0-146"},"data":{"baseType":"MetricData","baseData":{"ver":2,"metrics":[{"name":"Server response time","kind":"Aggregation","value":826.8681,"count":3,"min":257.0655,"max":289.3134,"stdDev":13.6074460079765}],"properties":{"operation/synthetic":"False","request/performanceBucket":"250ms-500ms","cloud/roleName":"Unknown","_MS.MetricId":"requests/duration","_MS.IsAutocollected":"True","_MS.AggregationIntervalMs":"50000","Request.Success":"True","cloud/roleInstance":"LAPTOP-XJEA31","request/resultCode":"200"}}}}
Application Insights Telemetry: {"name":"AppMetrics","time":"2021-06-29T13:19:11.0000000Z","iKey":"xxxxx","tags":{"ai.cloud.roleInstance":"LAPTOP-XJEA31","ai.internal.sdkVersion":"m-agg2:2.17.0-146"},"data":{"baseType":"MetricData","baseData":{"ver":2,"metrics":[{"name":"Dependency duration","kind":"Aggregation","value":427.3933,"count":1,"min":427.3933,"max":427.3933,"stdDev":0}],"properties":{"operation/synthetic":"False","Dependency.Success":"True","dependency/performanceBucket":"250ms-500ms","cloud/roleName":"Unknown","_MS.MetricId":"dependencies/duration","_MS.IsAutocollected":"True","dependency/target":"test.XXXX.api","_MS.AggregationIntervalMs":"50000","cloud/roleInstance":"LAPTOP-XJEA31","dependency/resultCode":"200","Dependency.Type":"Http"}}}}
Application Insights Telemetry: {"name":"AppMetrics","time":"2021-06-29T13:19:11.0000000Z","iKey":"xxxxx","tags":{"ai.cloud.roleInstance":"LAPTOP-XJEA31","ai.internal.sdkVersion":"m-agg2:2.17.0-146"},"data":{"baseType":"MetricData","baseData":{"ver":2,"metrics":[{"name":"Dependency duration","kind":"Aggregation","value":68.2655,"count":1,"min":68.2655,"max":68.2655,"stdDev":0}],"properties":{"operation/synthetic":"False","Dependency.Success":"True","dependency/performanceBucket":"<250ms","cloud/roleName":"Unknown","_MS.MetricId":"dependencies/duration","_MS.IsAutocollected":"True","dependency/target":"test.XXXX.api","_MS.AggregationIntervalMs":"50000","cloud/roleInstance":"LAPTOP-XJEA31","dependency/resultCode":"200","Dependency.Type":"Http"}}}}
Application Insights Telemetry: {"name":"AppMetrics","time":"2021-06-29T13:19:11.0000000Z","iKey":"xxxxx","tags":{"ai.cloud.roleInstance":"LAPTOP-XJEA31","ai.internal.sdkVersion":"m-agg2:2.17.0-146"},"data":{"baseType":"MetricData","baseData":{"ver":2,"metrics":[{"name":"Dependency duration","kind":"Aggregation","value":349.4797,"count":8,"min":26.1161,"max":78.3276,"stdDev":17.0806961506943}],"properties":{"operation/synthetic":"False","Dependency.Success":"True","dependency/performanceBucket":"<250ms","cloud/roleName":"Unknown","_MS.MetricId":"dependencies/duration","_MS.IsAutocollected":"True","dependency/target":"test.xxx.api","_MS.AggregationIntervalMs":"50000","cloud/roleInstance":"LAPTOP-XJEA31","dependency/resultCode":"200","Dependency.Type":"Http"}}}}
Application Insights Telemetry: {"name":"AppMetrics","time":"2021-06-29T13:19:11.0000000Z","iKey":"xxxxx","tags":{"ai.cloud.roleInstance":"LAPTOP-XJEA31","ai.internal.sdkVersion":"m-agg2:2.17.0-146"},"data":{"baseType":"MetricData","baseData":{"ver":2,"metrics":[{"name":"Dependency duration","kind":"Aggregation","value":3362.6157,"count":3,"min":1072.91,"max":1148.1796,"stdDev":34.0227879934419}],"properties":{"operation/synthetic":"False","Dependency.Success":"True","dependency/performanceBucket":"1sec-3sec","cloud/roleName":"Unknown","_MS.MetricId":"dependencies/duration","_MS.IsAutocollected":"True","dependency/target":"test.xxx.api","_MS.AggregationIntervalMs":"50000","cloud/roleInstance":"LAPTOP-XJEA31","dependency/resultCode":"200","Dependency.Type":"Http"}}}}
Application Insights Telemetry: {"name":"AppMetrics","time":"2021-06-29T13:19:11.0000000Z","iKey":"xxxxx","tags":{"ai.cloud.roleInstance":"LAPTOP-XJEA31","ai.internal.sdkVersion":"m-agg2:2.17.0-146"},"data":{"baseType":"MetricData","baseData":{"ver":2,"metrics":[{"name":"Server response time","kind":"Aggregation","value":3571.0884,"count":2,"min":1055.4554,"max":2515.633,"stdDev":730.0888}],"properties":{"operation/synthetic":"False","request/performanceBucket":"1sec-3sec","cloud/roleName":"Unknown","_MS.MetricId":"requests/duration","_MS.IsAutocollected":"True","_MS.AggregationIntervalMs":"50000","Request.Success":"True","cloud/roleInstance":"LAPTOP-XJEA31","request/resultCode":"200"}}}}
Application Insights Telemetry: {"name":"AppMetrics","time":"2021-06-29T13:19:11.0000000Z","iKey":"xxxxx","tags":{"ai.cloud.roleInstance":"LAPTOP-XJEA31","ai.internal.sdkVersion":"m-agg2:2.17.0-146"},"data":{"baseType":"MetricData","baseData":{"ver":2,"metrics":[{"name":"Dependency duration","kind":"Aggregation","value":1322.8034,"count":2,"min":603.9236,"max":718.8798,"stdDev":57.4780999999994}],"properties":{"operation/synthetic":"False","Dependency.Success":"True","dependency/performanceBucket":"500ms-1sec","cloud/roleName":"Unknown","_MS.MetricId":"dependencies/duration","_MS.IsAutocollected":"True","dependency/target":"test.xxx.api","_MS.AggregationIntervalMs":"50000","cloud/roleInstance":"LAPTOP-XJEA31","dependency/resultCode":"200","Dependency.Type":"Http"}}}}
Application Insights Telemetry: {"name":"AppMetrics","time":"2021-06-29T13:19:11.0000000Z","iKey":"xxxxx","tags":{"ai.cloud.roleInstance":"LAPTOP-XJEA31","ai.internal.sdkVersion":"m-agg2:2.17.0-146"},"data":{"baseType":"MetricData","baseData":{"ver":2,"metrics":[{"name":"Dependency duration","kind":"Aggregation","value":1306.2165,"count":3,"min":412.2061,"max":470.0995,"stdDev":24.9934099682295}],"properties":{"operation/synthetic":"False","Dependency.Success":"True","dependency/performanceBucket":"250ms-500ms","cloud/roleName":"Unknown","_MS.MetricId":"dependencies/duration","_MS.IsAutocollected":"True","dependency/target":"xxxxx.azurewebsites.net | cid-v1:6678289b-9f57-4f53-b932-9b94ba164c21","_MS.AggregationIntervalMs":"50000","cloud/roleInstance":"LAPTOP-XJEA31","dependency/resultCode":"200","Dependency.Type":"Http (tracked component)"}}}}
Application Insights Telemetry: {"name":"AppMetrics","time":"2021-06-29T13:19:11.0000000Z","iKey":"xxxxx","tags":{"ai.cloud.roleInstance":"LAPTOP-XJEA31","ai.internal.sdkVersion":"m-agg2:2.17.0-146"},"data":{"baseType":"MetricData","baseData":{"ver":2,"metrics":[{"name":"Server response time","kind":"Aggregation","value":0.0178,"count":2,"min":0.0084,"max":0.0094,"stdDev":0.000499999999999965}],"properties":{"operation/synthetic":"False","request/performanceBucket":"<250ms","cloud/roleName":"Unknown","_MS.MetricId":"requests/duration","_MS.IsAutocollected":"True","_MS.AggregationIntervalMs":"50000","Request.Success":"True","cloud/roleInstance":"LAPTOP-XJEA31","request/resultCode":"301"}}}}
Application Insights Telemetry: {"name":"AppMetrics","time":"2021-06-29T13:19:11.0000000Z","iKey":"xxxxx","tags":{"ai.cloud.roleInstance":"LAPTOP-XJEA31","ai.internal.sdkVersion":"m-agg2:2.17.0-146"},"data":{"baseType":"MetricData","baseData":{"ver":2,"metrics":[{"name":"Server response time","kind":"Aggregation","value":283.7352,"count":7,"min":20.0782,"max":91.4169,"stdDev":23.0686433140622}],"properties":{"operation/synthetic":"False","request/performanceBucket":"<250ms","cloud/roleName":"Unknown","_MS.MetricId":"requests/duration","_MS.IsAutocollected":"True","_MS.AggregationIntervalMs":"50000","Request.Success":"False","cloud/roleInstance":"LAPTOP-XJEA31","request/resultCode":"404"}}}}
Application Insights Telemetry: {"name":"AppMetrics","time":"2021-06-29T13:19:11.0000000Z","iKey":"xxxxx","tags":{"ai.cloud.roleInstance":"LAPTOP-XJEA31","ai.internal.sdkVersion":"m-agg2:2.17.0-146"},"data":{"baseType":"MetricData","baseData":{"ver":2,"metrics":[{"name":"Server response time","kind":"Aggregation","value":1560.7904,"count":37,"min":16.1433,"max":210.1531,"stdDev":34.1668982869377}],"properties":{"operation/synthetic":"False","request/performanceBucket":"<250ms","cloud/roleName":"Unknown","_MS.MetricId":"requests/duration","_MS.IsAutocollected":"True","_MS.AggregationIntervalMs":"50000","Request.Success":"True","cloud/roleInstance":"LAPTOP-XJEA31","request/resultCode":"200"}}}}
Application Insights Telemetry: {"name":"AppMetrics","time":"2021-06-29T13:19:11.0000000Z","iKey":"xxxxx","tags":{"ai.cloud.roleInstance":"LAPTOP-XJEA31","ai.internal.sdkVersion":"m-agg2:2.17.0-146"},"data":{"baseType":"MetricData","baseData":{"ver":2,"metrics":[{"name":"Dependency duration","kind":"Aggregation","value":1668.5255,"count":52,"min":15.4877,"max":86.7231,"stdDev":16.3124367529494}],"properties":{"operation/synthetic":"False","Dependency.Success":"True","dependency/performanceBucket":"<250ms","cloud/roleName":"Unknown","_MS.MetricId":"dependencies/duration","_MS.IsAutocollected":"True","dependency/target":"BencomStateServer.bencomfarm.local | ASPState","_MS.AggregationIntervalMs":"50000","cloud/roleInstance":"LAPTOP-XJEA31","dependency/resultCode":"Unknown","Dependency.Type":"SQL"}}}}
Exception thrown: 'System.UnauthorizedAccessException' in mscorlib.dll
Exception thrown: 'System.UnauthorizedAccessException' in mscorlib.dll
Exception thrown: 'System.UnauthorizedAccessException' in mscorlib.dll
Exception thrown: 'System.UnauthorizedAccessException' in mscorlib.dll
Exception thrown: 'System.UnauthorizedAccessException' in mscorlib.dll
Exception thrown: 'System.NullReferenceException' in Microsoft.AI.PerfCounterCollector.dll
Object reference not set to an instance of an object.`

**Comment by antoniogarcez** [comment]
Nope and it's really annoying.

To run my project, I have to click on "Continue" three times and after some time, the exception is thrown again and again and again... . If you find something, please tell me.

**Comment by Robert-Zandberg** [comment]
@antoniogarcez Can imagine. Do you have some other logging services installed? And does the originalString value also looks like this `"\\ASP.NET Applications(??APP_W3SVC_PROC??)\\Requests/Sec"` ?

**Comment by antoniogarcez** [comment]
@Robert-Zandberg Sorry but do you know where I can find that value? I am using the default logging services (ApplicationInsightsWebTracking)

**Comment by Robert-Zandberg** [comment]
@antoniogarcez I found the solution that works for me. Maybe this also helps you, the issue happened because the application pool does not have access to the folder where the website files are kept.  

So what I needed to do was adding a the IIS_IUSER as a authenticated user. Right click on the repo/folder then security tab, add new user and add IIS_IUSER. 
![image](https://user-images.githubusercontent.com/17293197/124268781-4361c880-db3a-11eb-9172-7b31b901c53b.png)

I hope this will help you as well.
